File: src/main/java/DataStructureAlgo/Java/LeetCode2025/ProblemSet/Arrays/_3152/SpecialArrayII_3152.py
"""
Author: Nitin Gupta
Date: 7/6/2025
Question Title: 3152. Special Array II
Link:  https://leetcode.com/problems/special-array-ii/description
"""

from collections import deque
from typing import List, Optional, Dict, Any

from helpers.common_methods import CommonMethods


class Solution:
    def isArraySpecial(self, nums: List[int], queries: List[List[int]]) -> List[bool]:
        """
        One way is to take queries range and perform the parity check just like 3151. Special Array I.
        However, it could possible that there will be many overlapping queries ranges.

        In order to find the parity of sub-array [start, end] which has length = end - start + 1
        we need to find the length of speicial array at index i. Means, we can maintain the length of longest special array seen so far till index i.

        For example:
        [4,3,1,6] then length special array at each index would be
        [1,1,1,2] since at index i=0 its length = 1
                  at index i = 2 its length = 1 as well since we have not seen so far any parity change
                  while at index i = 3 its length = 2 since just before 6, there is 1 with different parity

        similarly
        [1,2,3,4,5,6] => [1,2,3,4,5,6] since entire array is special array hence its Longest length is also same as array length.

        Now for every query, [start, end] we can simply check the length of longest special array at index end is equal to the length of query or not

        """

        # contains the length of longest special array at index [i]
        nums_length = len(nums)
        special_array = [1]

        # pre-process of special array length
        for i in range(1, nums_length):

            length = 1

            if nums[i - 1] % 2 != nums[i] % 2:
                length = special_array[i - 1] + 1

            special_array.append(length)

        # process query and build result
        result = []
        for query in queries:
            start = query[0]
            end = query[1]
            query_length = end - start + 1

            if query_length <= special_array[end]:
                result.append(True)
            else:
                result.append(False)

        return result


def test(input_data, queries, expected):
    """
    Test function to verify the solution
    """
    CommonMethods.print_test(["Input", "Expected"], True, input_data, expected)
    pass_test, final_pass = True, True
    output = Solution().isArraySpecial(input_data, queries)
    pass_test = CommonMethods.compare_result(output, expected, True)
    CommonMethods.print_test(["Output", "Pass"], False, output, "PASS" if pass_test else "FAIL")
    final_pass &= pass_test

    return final_pass


if __name__ == "__main__":
    tests: List = [test([3, 4, 1, 2, 6], [[0, 4]], [False]),
                   test([4, 3, 1, 6], [[0, 2], [2, 3]], [False, True]),
                   test([1, 8], [[1, 1]], [True]),
                   test([2, 7, 2], [[0, 0], [1, 2]], [True, True])]

    CommonMethods.print_all_test_results(tests)

File: src/main/java/DataStructureAlgo/Java/LeetCode2025/ProblemSet/Arrays/_3152/SpecialArrayII_3152.java
package DataStructureAlgo.Java.LeetCode2025.ProblemSet.Arrays._3152;

import DataStructureAlgo.Java.LeetCode2025.ProblemSet.DynamicProgramming._baseProblems.subsequence.LIS._300.LongestIncreasingSubsequence_300;
import DataStructureAlgo.Java.helpers.CommonMethods;

import java.util.*;


/**
 * Author: Nitin Gupta
 * Date: 09/12/24
 * Question Title: 3152. Special Array II
 * Link: https://leetcode.com/problems/special-array-ii/description
 * Description: An array is considered special if every pair of its adjacent elements contains two numbers with different parity.
 *
 * You are given an array of integer nums and a 2D integer matrix queries, where for queries[i] = [fromi, toi] your task is to check that
 * subarray
 *  nums[fromi..toi] is special or not.
 *
 * Return an array of booleans answer such that answer[i] is true if nums[fromi..toi] is special.
 *
 *
 *
 * Example 1:
 *
 * Input: nums = [3,4,1,2,6], queries = [[0,4]]
 *
 * Output: [false]
 *
 * Explanation:
 *
 * The subarray is [3,4,1,2,6]. 2 and 6 are both even.
 *
 * Example 2:
 *
 * Input: nums = [4,3,1,6], queries = [[0,2],[2,3]]
 *
 * Output: [false,true]
 *
 * Explanation:
 *
 * The subarray is [4,3,1]. 3 and 1 are both odd. So the answer to this query is false.
 * The subarray is [1,6]. There is only one pair: (1,6) and it contains numbers with different parity. So the answer to this query is true.
 *
 *
 * Constraints:
 *
 * 1 <= nums.length <= 105
 * 1 <= nums[i] <= 105
 * 1 <= queries.length <= 105
 * queries[i].length == 2
 * 0 <= queries[i][0] <= queries[i][1] <= nums.length - 1
 * <p><p>
 * File reference
 * -----------
 * Duplicate {@link}
 * Similar {@link SpecialArrayI_3151.py}
 * extension {@link LongestIncreasingSubsequence_300}
 * DP-BaseProblem {@link LongestIncreasingSubsequence_300}
 * <p><p>
 * Tags
 * -----
 * @medium
 * @Array
 * @BinarySearch
 * @PrefixSum
 * <p><p>
 * Company Tags
 * -----
 * <p>
 * -----
 * @Editorial <p><p>
 * -----
 * @OptimalSolution {@link }
 */
public class SpecialArrayII_3152 {

    public static void main(String[] args) {
        List<Boolean> tests = new ArrayList<>();
        tests.add(test(new int[]{3, 4, 1, 2, 6}, new int[][]{{0, 4}}, new boolean[]{false}));
        tests.add(test(new int[]{4, 3, 1, 6}, new int[][]{{0, 2}, {2, 3}}, new boolean[]{false, true}));
        tests.add(test(new int[]{2,7,2}, new int[][]{{0, 0}, {1, 2}}, new boolean[]{true, true}));
        tests.add(test(new int[]{1,8}, new int[][]{{1, 1}}, new boolean[]{true}));
        CommonMethods.printAllTestOutCome(tests);
    }

    private static boolean test(int[] nums, int [][]queries, boolean [] expected) {
        //add print here
        CommonMethods.printTest(new String[]{"Nums", "queries", "Expected"}, true, nums, queries, expected);

        boolean [] output;
        boolean pass, finalPass = true;

        Solution_DP solutionDp = new Solution_DP();
        output = solutionDp.isArraySpecial(nums, queries);
        pass = CommonMethods.compareResultOutCome(output, expected, true);
        finalPass = finalPass && pass;
        CommonMethods.printTest(new String[]{"DP", "Pass"}, false, output, pass ? "Pass" : "Fail");


        return finalPass;

    }

    static class Solution_DP {
        public boolean[] isArraySpecial(int[] nums, int[][] queries) {
            int n = nums.length;
            int []dp = new int[n]; //length of longest alternative sub-array of parity
            dp[0] = 1;

            for(int i = 1; i<n; i++){

                if((nums[i] & 1) != (nums[i-1]&1))
                    dp[i] = dp[i-1] + 1;
                else
                    dp[i] = 1;
            }

            boolean []result = new boolean[queries.length];
            int r = 0;

            for(int []q : queries){
                int start = q[0];
                int end = q[1];
                int len = end - start + 1;

                result[r++] = dp[end] >= len;
            }
            return result;
        }
    }
}
